### ISSUE TITLE
LoadError: UndefVarError: `NetCDFError` not defined

### ISSUE THREAD

**Issue by ryofurue** [Alexander-Barth/NCDatasets.jl#207]
**Describe the bug**
I get the error message in the subject line of this thread. Because my program doesn't use `NetCDFError` directly, the error must be from within the `NCDatasets` package:
```julia
ERROR: LoadError: UndefVarError: `NetCDFError` not defined
```
[Full stack trace at the bottom of this message.]

**To Reproduce**
[I hope that the above description is enough to direct you to the right place in your code, but if you developers don't find the error obvious, I'll try to create a minimal example code to reproduce the error.]

**Expected behavior**
N/A

**Environment**

 - operating system: macOS 13.4
 - Julia version: 1.9.1 from `juliaup`
 - Output of the julia command `versioninfo()`
     Julia Version 1.9.1
    Commit 147bdf428cd (2023-06-07 08:27 UTC)
    Platform Info:
      OS: macOS (arm64-apple-darwin22.4.0)
      CPU: 10 × Apple M2 Pro
      WORD_SIZE: 64
      LIBM: libopenlibm
      LLVM: libLLVM-14.0.6 (ORCJIT, apple-m1)
      Threads: 1 on 6 virtual cores
 - NCDatasets version: v0.12.16
 - Output of `using Pkg; Pkg.status(mode=PKGMODE_MANIFEST)`
Status `~/.julia/environments/v1.9/Manifest.toml`
  [d1d4a3ce] BitFlags v0.1.7
  [179af706] CFTime v0.1.2
  [944b1d66] CodecZlib v0.7.1
  [35d6a980] ColorSchemes v3.21.0
  [3da002f7] ColorTypes v0.11.4
  [c3611d14] ColorVectorSpace v0.9.10
  [5ae59095] Colors v0.12.10
  [1fbeeb36] CommonDataModel v0.2.2
⌅ [34da2185] Compat v3.46.2
  [f0e56b4a] ConcurrentUtilities v2.2.0
  [187b0558] ConstructionBase v1.5.2
  [d38c429a] Contour v0.6.2
  [9a962f9c] DataAPI v1.15.0
  [864edb3b] DataStructures v0.18.13
  [8bb1440f] DelimitedFiles v1.9.1
  [ffbed154] DocStringExtensions v0.9.3
  [c87230d0] FFMPEG v0.4.1
  [53c48c17] FixedPointNumbers v0.8.4
  [59287772] Formatting v0.4.2
  [28b8d3ca] GR v0.72.7
  [9a22fb26] GibbsSeaWater v0.1.2
  [42e2da0e] Grisu v1.0.2
  [cd3eb016] HTTP v1.9.6
  [92d709cd] IrrationalConstants v0.2.2
  [1019f520] JLFzf v0.1.5
  [692b3bcd] JLLWrappers v1.4.1
  [682c06a0] JSON v0.21.4
  [b964fa9f] LaTeXStrings v1.3.0
  [23fbe1c1] Latexify v0.16.0
  [2ab3a3ac] LogExpFunctions v0.3.24
  [e6f89c97] LoggingExtras v1.0.0
  [1914dd2f] MacroTools v0.5.10
  [739be429] MbedTLS v1.1.7
  [442fdcdd] Measures v0.3.2
  [e1d29d7a] Missings v1.1.0
  [85f8d34a] NCDatasets v0.12.16
  [77ba4419] NaNMath v1.0.2
  [4d8831e6] OpenSSL v1.4.1
  [bac558e1] OrderedCollections v1.6.0
  [69de0a69] Parsers v2.5.10
  [b98c9c47] Pipe v1.3.0
  [ccf2f8ad] PlotThemes v3.1.0
  [995b91a9] PlotUtils v1.3.5
  [91a5bcdd] Plots v1.38.15
  [aea7be01] PrecompileTools v1.1.2
  [21216c6a] Preferences v1.4.0
  [3cdcf5f2] RecipesBase v1.3.4
  [01d81517] RecipesPipeline v0.6.12
  [189a3867] Reexport v1.2.2
  [05181044] RelocatableFolders v1.0.0
  [ae029012] Requires v1.3.0
  [6c6a2e73] Scratch v1.2.0
  [e8f3a9d7] SearchSortedNearest v0.1.1
  [992d4aef] Showoff v1.0.3
  [777ac1f9] SimpleBufferStream v1.1.0
  [a2af1166] SortingAlgorithms v1.1.0
  [276daf66] SpecialFunctions v2.2.0
  [82ae8749] StatsAPI v1.6.0
  [2913bbd2] StatsBase v0.34.0
  [62fd8b95] TensorCore v0.1.1
  [3bb67fe8] TranscodingStreams v0.9.13
  [5c2747f8] URIs v1.4.2
  [1cfade01] UnicodeFun v0.4.1
  [1986cc42] Unitful v1.14.0
  [45397f5d] UnitfulLatexify v1.6.3
  [41fe7b60] Unzip v0.2.0
  [6e34b625] Bzip2_jll v1.0.8+0
  [83423d85] Cairo_jll v1.16.1+1
  [2e619515] Expat_jll v2.4.8+0
  [b22a6f82] FFMPEG_jll v4.4.2+2
  [a3f928ae] Fontconfig_jll v2.13.93+0
  [d7e528f0] FreeType2_jll v2.10.4+0
  [559328eb] FriBidi_jll v1.0.10+0
  [0656b61e] GLFW_jll v3.3.8+0
  [d2c73de3] GR_jll v0.72.7+0
  [78b55507] Gettext_jll v0.21.0+0
  [6727f6b2] GibbsSeaWater_jll v3.5.2+0
  [7746bdde] Glib_jll v2.74.0+2
  [3b182d85] Graphite2_jll v1.3.14+0
⌅ [0234f1f7] HDF5_jll v1.12.2+2
  [2e76f6c2] HarfBuzz_jll v2.8.1+1
  [aacddb02] JpegTurbo_jll v2.1.91+0
  [c1c5ebd0] LAME_jll v3.100.1+0
  [88015f11] LERC_jll v3.0.0+1
  [dd4b983a] LZO_jll v2.10.1+0
⌅ [e9f186c6] Libffi_jll v3.2.2+1
  [d4300ac3] Libgcrypt_jll v1.8.7+0
  [7e76a0d4] Libglvnd_jll v1.6.0+0
  [7add5ba3] Libgpg_error_jll v1.42.0+0
  [94ce4f54] Libiconv_jll v1.16.1+2
  [4b2f31a3] Libmount_jll v2.35.0+0
  [89763e89] Libtiff_jll v4.4.0+0
  [38a345b3] Libuuid_jll v2.36.0+0
⌃ [7243133f] NetCDF_jll v400.902.5+1
  [e7412a2a] Ogg_jll v1.3.5+1
⌅ [458c3c95] OpenSSL_jll v1.1.21+0
  [efe28fd5] OpenSpecFun_jll v0.5.5+0
  [91d4177d] Opus_jll v1.3.2+0
  [30392449] Pixman_jll v0.40.1+0
  [ea2cea3b] Qt5Base_jll v5.15.3+2
  [a2964d1f] Wayland_jll v1.21.0+0
  [2381bf8a] Wayland_protocols_jll v1.25.0+0
  [02c8fc9c] XML2_jll v2.10.3+0
  [aed1982a] XSLT_jll v1.1.34+0
  [4f6342f7] Xorg_libX11_jll v1.6.9+4
  [0c0b7dd1] Xorg_libXau_jll v1.0.9+4
  [935fb764] Xorg_libXcursor_jll v1.2.0+4
  [a3789734] Xorg_libXdmcp_jll v1.1.3+4
  [1082639a] Xorg_libXext_jll v1.3.4+4
  [d091e8ba] Xorg_libXfixes_jll v5.0.3+4
  [a51aa0fd] Xorg_libXi_jll v1.7.10+4
  [d1454406] Xorg_libXinerama_jll v1.1.4+4
  [ec84b674] Xorg_libXrandr_jll v1.5.2+4
  [ea2f1a96] Xorg_libXrender_jll v0.9.10+4
  [14d82f49] Xorg_libpthread_stubs_jll v0.1.0+3
  [c7cfdc94] Xorg_libxcb_jll v1.13.0+3
  [cc61e674] Xorg_libxkbfile_jll v1.1.0+4
  [12413925] Xorg_xcb_util_image_jll v0.4.0+1
  [2def613f] Xorg_xcb_util_jll v0.4.0+1
  [975044d2] Xorg_xcb_util_keysyms_jll v0.4.0+1
  [0d47668e] Xorg_xcb_util_renderutil_jll v0.3.9+1
  [c22f9ab0] Xorg_xcb_util_wm_jll v0.4.1+1
  [35661453] Xorg_xkbcomp_jll v1.4.2+4
  [33bec58e] Xorg_xkeyboard_config_jll v2.27.0+4
  [c5fb5394] Xorg_xtrans_jll v1.4.0+3
  [3161d3a3] Zstd_jll v1.5.5+0
⌅ [214eeab7] fzf_jll v0.29.0+0
  [a4ae2306] libaom_jll v3.4.0+0
  [0ac62f75] libass_jll v0.15.1+0
  [f638f0a6] libfdk_aac_jll v2.0.2+0
  [b53b4c65] libpng_jll v1.6.38+0
  [f27f6e37] libvorbis_jll v1.3.7+1
  [1270edf5] x264_jll v2021.5.5+0
  [dfaa095f] x265_jll v3.5.0+0
  [d8fb68d0] xkbcommon_jll v1.4.1+0
  [0dad84c5] ArgTools v1.1.1
  [56f22d72] Artifacts
  [2a0f44e3] Base64
  [ade2ca70] Dates
  [8ba89e20] Distributed
  [f43a241f] Downloads v1.6.0
  [7b1f6079] FileWatching
  [b77e0a4c] InteractiveUtils
  [b27032c2] LibCURL v0.6.3
  [76f85450] LibGit2
  [8f399da3] Libdl
  [37e2e46d] LinearAlgebra
  [56ddb016] Logging
  [d6f4376e] Markdown
  [a63ad114] Mmap
  [ca575930] NetworkOptions v1.2.0
  [44cfe95a] Pkg v1.9.0
  [de0858da] Printf
  [3fa0cd96] REPL
  [9a3f8284] Random
  [ea8e919c] SHA v0.7.0
  [9e88b42a] Serialization
  [1a1011a3] SharedArrays
  [6462fe0b] Sockets
  [2f01184e] SparseArrays
  [10745b16] Statistics v1.9.0
  [fa267f1f] TOML v1.0.3
  [a4e569a6] Tar v1.10.0
  [8dfed614] Test
  [cf7118a7] UUIDs
  [4ec0a83e] Unicode
  [e66e0078] CompilerSupportLibraries_jll v1.0.2+0
  [deac9b47] LibCURL_jll v7.84.0+0
  [29816b5a] LibSSH2_jll v1.10.2+0
  [c8ffd9c3] MbedTLS_jll v2.28.2+0
  [14a3606d] MozillaCACerts_jll v2022.10.11
  [4536629a] OpenBLAS_jll v0.3.21+4
  [05823500] OpenLibm_jll v0.8.1+0
  [efcefdf7] PCRE2_jll v10.42.0+0
  [bea87d4a] SuiteSparse_jll v5.10.1+6
  [83775a58] Zlib_jll v1.2.13+0
  [8e850b90] libblastrampoline_jll v5.8.0+0
  [8e850ede] nghttp2_jll v1.48.0+0
  [3f19e933] p7zip_jll v17.4.0+0
Info Packages marked with ⌃ and ⌅ have new versions available, but those with ⌅ are restricted by compatibility constraints from upgrading. To see why use `status --outdated -m` 

**Full output**

In case of an error, please paste the full error message and stack trace.
```
ERROR: LoadError: UndefVarError: `NetCDFError` not defined
Stacktrace:
 [1] setindex!(v::CommonDataModel.CFVariable{Float64, 1, NCDatasets.Variable{Float64, 1, NCDataset{Nothing}}, NCDatasets.Attributes{NCDataset{Nothing}}, NamedTuple{(:fillvalue, :missing_values, :scale_factor, :add_offset, :calendar, :time_origin, :time_factor), Tuple{Nothing, Tuple{}, Vararg{Nothing, 5}}}}, data::Vector{DateTime}, indexes::Colon)
   @ CommonDataModel ~/.julia/packages/CommonDataModel/7hLGg/src/cfvariable.jl:412
 [2] (::var"#18#19"{Vector{Union{Missing, Float64}}, String, Vector{DateTime}, Vector{Float64}})(ds::NCDataset{Nothing})
   @ Main ~/Dropbox/work/trenches/bottom_pressure-03.jl:305
 [3] NCDataset(::var"#18#19"{Vector{Union{Missing, Float64}}, String, Vector{DateTime}, Vector{Float64}}, ::String, ::Vararg{String}; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ NCDatasets ~/.julia/packages/NCDatasets/wsGwu/src/dataset.jl:255
 [4] NCDataset
   @ ~/.julia/packages/NCDatasets/wsGwu/src/dataset.jl:252 [inlined]
 [5] calc_save_pres_1d(; sta::ObsStation, slp::Vector{Union{Missing, Float64}}, em::Vector{Union{Missing, Float64}}, sm::Matrix{Union{Missing, Float64}}, tm::Matrix{Union{Missing, Float64}}, delz::Vector{Float64}, lon::Float64, lat::Float64, tax::Tuple{Vector{DateTime}, String})
   @ Main ~/Dropbox/work/trenches/bottom_pressure-03.jl:297
 [6] process_sta(sta::ObsStation)
   @ Main ~/Dropbox/work/trenches/bottom_pressure-03.jl:353
 [7] top-level scope
   @ ~/Dropbox/work/trenches/bottom_pressure-03.jl:400
 [8] include(fname::String)
   @ Base.MainInclude ./client.jl:478
 [9] top-level scope
   @ REPL[1]:1
in expression starting at /Users/furue/Dropbox/work/trenches/bottom_pressure-03.jl:400
```

**Comment by Alexander-Barth** [comment]
I believe that this is an issue with CommonDataModel.jl 0.2.2 which should be fixed by the new released 0.2.3:

https://github.com/JuliaRegistries/General/pull/85186

Note however, that after updating you will see this error message:
https://github.com/JuliaGeo/CommonDataModel.jl/blob/v0.2.2/src/cfvariable.jl#L412


It seems that you are changing the time units or calendar attribute.  Is this fixed if you use this instead?
https://alexander-barth.github.io/NCDatasets.jl/dev/issues/#Defining-the-attributes-_FillValue,-add_offset,-scale_factor,-units-and-calendar

**Comment by ryofurue** [comment]
> It seems that you are changing the time units or calendar attribute

Close.  After I submitted this report, I found the cause of the error: I was trying to assign an array of `DateTime`s to the time axis, which is defined to be an array of `Float64`s:
```julia
taxis = defVar(ds,"time",Float64,("time",))
taxis[:] = tvals[:] # array of DateTime
```

Incidentally, how do you use the array of `DateTime`s for the time axis?  This time axis is actually read from another netCDF file.  I just want to use the same axis for the new netCDF file I'm creating.

**Comment by Alexander-Barth** [comment]
something like this?

```julia
other_ds = NCDataset("other_file.nc");
taxis = defVar(ds,"time",Float64,("time",),attrib = other_ds["time"].attrib)
taxis[:] = other_ds["time"][:]
```

maybe we can have something like this in future to make this simpler:


```julia
other_ds = NCDataset("other_file.nc");
taxis = defVar(ds,other_ds["time"]) # copies all data and attributes under the same name (to be implemented)
```

**Comment by ryofurue** [comment]
```julia
taxis = defVar(ds,"time",Float64, . . .)
taxis[:] = other_ds["time"][:]
```
It's quite surprising! (I'm not against it. I'm just surprised.)  that you can assign`DateTime` objects to `taxis[:]`. 
Because `taxis[:]` behaves like a vector of `Float64`, I thought that assignment wouldn't work.

> maybe we can have something like this in future to make this simpler:
> ```julia
> other_ds = NCDataset("other_file.nc");
> taxis = defVar(ds,other_ds["time"]) # copies all data and attributes under the same name (to be implemented)
> ```

That would be definitely nicer!

Here is a concrete scenario I sometimes have to deal with.  Suppose you have a timeseries defined at the beginning of each hour.  Now you want to shift the time axis by 30 minutes (because the time axis of the original dataset is wrong):
```
 newtaxis = defVar(ds,other_ds["time"])
 newtaxis[:] = newtaxis[:] + Dates.Minute(30)
```
and you don't have to worry about the attributes of the axis.

It would be nice if the time origin attribute accepts a DateTime object.

**Comment by Alexander-Barth** [comment]
This code should work now with the master version (of NCDatasets and CommonDataModel).

```julia
newtaxis = defVar(ds,other_ds["time"])
newtaxis[:] = newtaxis[:] .+ Dates.Minute(30)
```

There is a test case here:
https://github.com/Alexander-Barth/NCDatasets.jl/blob/master/test/test_variable.jl#L263


Thank you for your feedback. Can this issue be closed?

**Comment by ryofurue** [comment]
> Can this issue be closed?

I thought the original issue has long been closed.  Specifically, the issue my original report pointed out was closed by your comment "I believe that this is an issue with CommonDataModel.jl 0.2.2 which should be fixed by the new released 0.2.3" more than a week ago.

After that, we have been discussing a potential enhancement to the NCDatasets package.  That is, we have been "abusing" this thread :-)

Okay, I'll press the "Close with comment" button now, if I have the right to do so.

**Comment by tcarion** [comment]
> (for your information: defVar should also work between NetCDF and GRIB files; so you can copy a variable in a GRIB file and write it with metadata to a NetCDF file, @tcarion )

Another reason why `CommonDataModel` was definitely worth it :-)
